File: androidApp/src/main/java/bisq/client/android/presentation/notification_detail/NotificationDetailViewModel.kt
package bisq.client.android.presentation.notification_detail

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import bisq.client.domain.model.GenericMessageInfo
import bisq.client.domain.util.GenericMessageInfoQueueUtil
import bisq.client.domain.util.Queue
import bisq.client.interactors.notification_detail.GetNotification
import bisq.client.presentation.notification_detail.NotificationDetailEvents
import bisq.client.presentation.notification_detail.NotificationDetailState
import bisq.client.util.Logger
import bisq.client.domain.model.UIComponentType
import bisq.client.interactors.notification_detail.RemoveNotification
import bisq.client.presentation.notification_list.NotificationListEvents
import dagger.hilt.android.lifecycle.HiltViewModel
import java.util.*
import javax.inject.Inject

@HiltViewModel
class NotificationDetailViewModel
@Inject
constructor(
    private val savedStateHandle: SavedStateHandle,
    private val getNotification: GetNotification,
    private val removeNotification: RemoveNotification
): ViewModel() {

    private val logger = Logger("NotificationDetailViewModel")

    val state: MutableState<NotificationDetailState> = mutableStateOf(NotificationDetailState())

    init {
        savedStateHandle.get<Int>("notificationId")?.let { notificationId ->
            onTriggerEvent(NotificationDetailEvents.GetNotification(notificationId = notificationId))
        }
    }

    fun onTriggerEvent(event: NotificationDetailEvents) {
        when (event) {
            is NotificationDetailEvents.GetNotification -> {
                getNotification(notificationId = event.notificationId)
            }
            is NotificationDetailEvents.RemoveNotification -> {
                removeNotification(notificationId = event.notificationId)
            }
            is NotificationDetailEvents.OnRemoveHeadMessageFromQueue -> {
                removeHeadMessage()
            }
            else -> {
                val messageInfoBuilder = GenericMessageInfo.Builder()
                    .id(UUID.randomUUID().toString())
                    .title("Invalid Event")
                    .uiComponentType(UIComponentType.Dialog)
                    .description("Something went wrong.")
                appendToMessageQueue(messageInfo = messageInfoBuilder)
            }
        }
    }

    private fun removeHeadMessage() {
        try {
            val queue = state.value.queue
            queue.remove() // can throw exception if empty
            state.value = state.value.copy(queue = Queue(mutableListOf())) // force recompose
            state.value = state.value.copy(queue = queue)
        } catch (e: Exception) {
            logger.log("Nothing to remove from DialogQueue")
        }
    }

    private fun getNotification(notificationId: Int) {
        getNotification.execute(notificationId = notificationId).collectCommon(viewModelScope) { dataState ->
            state.value = state.value.copy(isLoading = dataState.isLoading)

            dataState.data?.let { notification ->
                state.value = state.value.copy(notification = notification)
            }

            dataState.message?.let { message ->
                appendToMessageQueue(message)
            }
        }
    }

    private fun removeNotification(notificationId: Int) {
        removeNotification.execute(notificationId = notificationId).collectCommon(viewModelScope) { dataState ->
            state.value = state.value.copy(isLoading = dataState.isLoading)

            dataState.message?.let { message ->
                appendToMessageQueue(message)
            }
        }
    }

    private fun appendToMessageQueue(messageInfo: GenericMessageInfo.Builder) {
        if (!GenericMessageInfoQueueUtil()
                .doesMessageAlreadyExistInQueue(queue = state.value.queue,
                    messageInfo = messageInfo.build())) {
            val queue = state.value.queue
            queue.add(messageInfo.build())
            state.value = state.value.copy(queue = queue)
        }
    }
}

File: iosApp/iosApp/presentation/notification_detail/NotificationDetailViewModel.swift
import SwiftUI
import common

class NotificationDetailViewModel: ObservableObject {
    
    private let logger = Logger(className: "NotificationDetailViewModel")

    private let getNotification: GetNotification

    @Published var state: NotificationDetailState = NotificationDetailState()
    
    init(
        notificationId: Int,
        getNotification: GetNotification
    ) {
        self.getNotification = getNotification
        onTriggerEvent(stateEvent: NotificationDetailEvents.GetNotification(notificationId: Int32(notificationId)))
    }
    
    func onTriggerEvent(stateEvent: NotificationDetailEvents) {
        switch stateEvent {
        case is NotificationDetailEvents.GetNotification:
            getNotification(notificationId: Int((stateEvent as! NotificationDetailEvents.GetNotification).notificationId))
        case is NotificationDetailEvents.OnRemoveHeadMessageFromQueue:
            removeHeadFromQueue()
        default: doNothing()
        }
    }
        
    private func getNotification(notificationId: Int) {
        do {
            try self.getNotification.execute(
                notificationId: Int32(notificationId)
            ).collectCommon(
                coroutineScope: nil,
                callback: { dataState in
                if (dataState != nil) {
                    let data = dataState?.data
                    let message = dataState?.message
                    let loading = dataState?.isLoading ?? false
                    self.updateState(isLoading: loading)
                    
                    if (data != nil) {
                        self.updateState(notification: data! as common.Notification)
                    }
                    if (message != nil) {
                        self.handleMessageByUIComponentType(message!.build())
                    }
                } else {
                    self.logger.log(msg: "GetNotification: DataState is nil")
                }
            })
        } catch {
            self.logger.log(msg: "\(error)")
        }
    }
    
    private func handleMessageByUIComponentType(_ message: GenericMessageInfo) {
        switch message.uiComponentType {
        case UIComponentType.Dialog():
            appendToQueue(message: message)
        case UIComponentType.None():
            logger.log(msg: "\(message.description)")
        default:
            doNothing()
        }
    }
    
    private func appendToQueue(message: GenericMessageInfo) {
        let currentState = (self.state.copy() as! NotificationDetailState)
        let queue = currentState.queue
        let queueUtil = GenericMessageInfoQueueUtil() // prevent duplicates
        if !queueUtil.doesMessageAlreadyExistInQueue(queue: queue, messageInfo: message) {
           queue.add(element: message)
           updateState(queue: queue)
        }
   }
    
    /**
     *  Remove the head message from queue
     */
    private func removeHeadFromQueue() {
        let currentState = (self.state.copy() as! NotificationDetailState)
        let queue = currentState.queue
        do {
            try queue.remove()
            updateState(queue: queue)
        } catch {
            logger.log(msg: "\(error)")
        }
    }
    
    private func doNothing() {}
    
    private func updateState(
        isLoading: Bool? = nil,
        notification: common.Notification? = nil,
        queue: Queue<GenericMessageInfo>? = nil
    ) {
        let currentState = (self.state.copy() as! NotificationDetailState)
        self.state = self.state.doCopy(
            isLoading: isLoading ?? currentState.isLoading,
            notification: notification ?? currentState.notification,
            queue: queue ?? currentState.queue
        )
    }
    
}
